pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java

```diff
@@ -172,48 +170,55 @@ private IntermediateResultsBlock combineBlocks()
         public void runJob() {
           AggregationGroupByResult groupByResult;
 
-          // If block is null (unexpected error), or there's no group-by result, then return as there's nothing to do.
-          blocks[index] = (IntermediateResultsBlock) _operators.get(index).nextBlock();
-          if (blocks[index] != null) {
-            groupByResult = blocks[index].getAggregationGroupByResult();
-
-            if (groupByResult != null) {
-              // Iterate over the group-by keys, for each key, update the group-by result in the resultsMap.
-              Iterator<GroupKeyGenerator.GroupKey> groupKeyIterator = groupByResult.getGroupKeyIterator();
-
-              while (groupKeyIterator.hasNext()) {
-                GroupKeyGenerator.GroupKey groupKey = groupKeyIterator.next();
-                String groupKeyString = groupKey.getStringKey();
-
-                // HashCode method might return negative value, make it non-negative
-                int lockIndex = (groupKeyString.hashCode() & Integer.MAX_VALUE) % NUM_LOCKS;
-                synchronized (LOCKS[lockIndex]) {
-                  Serializable[] results = resultsMap.get(groupKeyString);
-
-                  if (results == null) {
-                    results = new Serializable[numAggrFunctions];
-                    for (int j = 0; j < numAggrFunctions; j++) {
-                      results[j] = groupByResult.getResultForKey(groupKey, j);
-                    }
-
-                    resultsMap.put(groupKeyString, results);
-                  } else {
-                    for (int j = 0; j < numAggrFunctions; j++) {
-                      results[j] = aggregationFunctions.get(j)
-                          .combineTwoValues(results[j], groupByResult.getResultForKey(groupKey, j));
+          try {
+            // If block is null (unexpected error), or there's no group-by result, then return as there's nothing to do.
+            blocks[index] = (IntermediateResultsBlock) _operators.get(index).nextBlock();
```

Comment:
The check for blocks[index] being null was already there. (Line 177)  I have seen the NPE on the pinot_server in the case of a timeout.